### DISCUSSION TITLE
Setting ligature caret position in Robofont?

### DISCUSSION THREAD

**Discussion by sponcey** [fonttools/fontbakery#3019] [Q&A]
This warn of the ligature caret position gives instructions of how to do this in Glyphs. Are there any open source examples of how to do this with projects built from .ufo's/Robofont? Maybe @thundernixon knows about this? ;)

> ⚠ WARN: Are there caret positions declared for every ligature?
> com.google.fonts/check/ligature_carets
> --- Rationale ---
> 
> All ligatures in a font must have corresponding caret (text cursor) positions
> defined in the GDEF table, otherwhise, users may experience issues with caret
> rendering.
> 
> If using GlyphsApp, ligature carets can be set directly on canvas by accessing
> the `Glyph -> Set Anchors` menu option or by pressing the `Cmd+U` keyboard
> shortcut.
> 
> 
> ⚠ WARN GDEF table is missing, but it is mandatory to declare it on fonts that provide ligature glyphs because the caret (text cursor) positioning for each ligature must be provided in this table. [code: GDEF-missing]

https://github.com/undercasetype/Fraunces/issues/220

**Comment by felipesanches** [comment]
Thanks for posting this issue, @sponcey! I fully agree that FontBakery should include tips for multiple font editors. I dislike making it GlyphsApp-biased, especially since Glyphs is proprietary. The reason we included this was because it is one of the most commonly used editors, even for libre fonts, which is a cultural contradiction from the point of view of the software freedom movement (which I am strongly aligned to), but it's a sad reality of the type-design industry nowadays.

At least glyphs files can be perfectly compiled into font binaries with a fully-free toolchain (**fontmake** and auxiliary post-processing scripts like the ones offered by **gftools**).

At first, I'd add multiple instructions to check rationales, but later we could consider auto-detecting which font editor was used on a given project and then give the user the appropriate instructions.

**Comment by arrowtype** [comment]
Updating this:

This is now part of the UFO spec: https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#ligature-carets

> Anchor names starting with caret_ provide the x value for a ligature caret position, and the y value is ignored. Likewise, anchor names starting with vcaret_ provide the y value for a ligature caret position, and the x value is ignored.

But, ufo2ft does not yet build these into fonts. [As of October 2020](https://github.com/googlefonts/fontbakery/pull/3063#issuecomment-709875116), @anthrotype said:

> The "caret_X" are supported by fontmake, but only when generating straight from .glyphs sources (because glyphsLib writes them in the GDEF table definition in the UFO features.fea). When generating from pure designs pace+UFOs (i.e. not in turn created by glyphsLib), ufo2ft doesn't understand this kind of anchors yet. But it will compile ligature carets if these are defined in the features.fea GDEF table.

I’m not currently sure how these would be defined in the features.fea GDEF table.

**Reply by madig** [reply]
ufo2ft now generates the GDEF with ligature positions if your features.fea does not already contain a GDEF.
